Firestore/core/src/firebase/firestore/model/mutations.cc

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/mutations.h"
+
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/model/document.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/model/precondition.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+MutationResult::MutationResult(
+    absl::optional<SnapshotVersion> version,
+    absl::optional<std::vector<FieldValue>> transform_results)
+    : version_(std::move(version)),
+      transform_results_(std::move(transform_results)) {
+}
+
+Mutation::Mutation(DocumentKey key, Precondition precondition)
+    : key_(std::move(key)), precondition_(std::move(precondition)) {
+}
+
+SetMutation::SetMutation(DocumentKey key,
+                         FieldValue value,
+                         Precondition precondition)
+    : Mutation(std::move(key), std::move(precondition)),
+      value_(std::move(value)) {
+}
+
+MaybeDocumentPointer SetMutation::ApplyTo(
+    const MaybeDocumentPointer& maybe_doc,
+    const MaybeDocumentPointer& /* base_doc */,
+    const Timestamp& /* local_write_time */,
+    const absl::optional<MutationResult>& mutation_result) const {
+  if (mutation_result) {
+    FIREBASE_ASSERT_MESSAGE(!mutation_result->transform_results(),
+                            "Transform results received by SetMutation.");
+  }
+
+  if (!precondition_.IsValidFor(maybe_doc)) {
```

Comment:
Ack. Just my one cent, I'd like to not over-generalizing i.e. allowing some duplicated code and meanwhile keep the interface flexible i.e. subtypes can do anything by overriding `ApplyTo` instead of making any `Mutation` must following a certain logic structure.